books/projects/x86isa/machine/instructions/bit.lisp

```diff
@@ -822,3 +822,85 @@
     x86))
 
 ;; ======================================================================
+
+
+(local (in-theory (disable bitops::ash-1-removal
+                           signed-byte-p
+                           unsigned-byte-p)))
+
+(define blsi ((operand-size posp)
+              (src-val natp))
+  (b* ((bitwidth (ash operand-size 3))
+       (tzcnt (tzcnt bitwidth 0 src-val)))
+    (loghead bitwidth (ash 1 tzcnt)))
+  ///
+  (defthm unsigned-byte-p-of-blsi
+    (implies (and (integerp w)
+                  (posp operand-size)
+                  (<= (* 8 operand-size) w))
+             (unsigned-byte-p w (blsi operand-size src-val)))
+    :hints(("Goal" :in-theory (enable bitops::ash-is-expt-*-x)))))
+       
+
+(def-inst x86-blsi
+  ;; VEX.LZ.0F38.W0 F3 /3:  BLSI r32, r/m32
+  ;; VEX.LZ.0F38.W1 F3 /3:  BLSI r64, r/m64
+  :parents (three-byte-opcodes)
+
+  :returns (x86 x86p :hyp (x86p x86))
+  :guard (vex-prefixes-byte0-p vex-prefixes)
+
+  :modr/m t
+  :vex t
+  :guard-debug t
+  :body
+  ;; placeholder
+  (b* ((operand-size (if (eql (vex->w vex-prefixes) 1) 8 4))
```

Comment:
The SDM says VEX.W1 is ignored in non-64 bit modes. Should that be checked here? I'm not very familiar with how 32-bit support works in the model.

Reply:
Thanks, good catch.